### PR TITLE
Various OpInfo architecture improvements

### PR DESCRIPTION
**Pull request by mruberry**
This PR makes the following improvements:

- moves the custom skip list for test_normalize_operator_exhaustive in test_fx_experimental to use the typical OpInfo skip architecture. The skips were updated to xfails, and that identified some operators which were no longer failing the test
- redundant tests with OpInfo-based testing in test_jit.py were removed
- test_dtypes was improved so its error messages are clear and it makes test_nondifferentiable redundant; the latter test has been removed
- OpInfo.supports_complex_autograd() is removed in favor of a more accurate and general test for whether the particular dtype is in the backward dtypes of the operator
- gradchecks have been improved to verify that an operator doesn't support grad if it claims not to
- gradchecks have been improved to test the gradient of all input tensors that require gradient
- the concept of "default test dtypes" has been removed
- excessive and mostly redundant out testing for elementwise unary operators has been removed
- metadata for whether an op supports nuanced "safe casting" to out behavior has been removed from OpInfos
- numerous skips have been converted to xfails
- numerous OpInfos have had their metadata fixed based on the new checks
- jit-specific utilities in common_methods_invocations.py have been moved to jit_programming_utils.py

### CODE REVIEW

File: test/test_unary_ufuncs.py [hunk 1]
```diff
@@ -486,34 +489,6 @@ def test_batch_vs_slicing(self, device, dtype, op):
...
-        if expect_fail:
-            with self.assertRaises(RuntimeError):
-                op(input, out=output, **kwargs)
-        else:
-            res = op(input, out=output, **kwargs)
-            self.assertTrue(res is output)
-            self.assertEqual(output, expected.to(output.dtype))
-
-    @ops(unary_ufuncs, dtypes=OpDTypes.supported)
-    def test_out_arg_all_dtypes(self, device, dtype, op):
```
Comment (ngimel) [inline test/test_unary_ufuncs.py:504]: is this checked in test_out? Not for all dtypes, I'm sure, but that's probably fine

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -9560,14 +9530,20 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
            supports_forward_ad=True,
            supports_fwgrad_bwgrad=True,
            skips=(
+               # Float did not match double
+               DecorateInfo(unittest.expectedFailure, 'TestGradients', 'test_fn_grad'),
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:9534]: so gradients are all wrong here?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -14152,14 +14023,14 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
...
                    dtypesIfCUDA=all_types_and(torch.bool, torch.half),
-                   safe_casts_outputs=True,
                    supports_forward_ad=True,
                    supports_fwgrad_bwgrad=True,
                    sample_inputs_func=sample_inputs_polygamma,
                    skips=(
                        # Redundant tests
                        DecorateInfo(unittest.skip("Skipped!"), 'TestGradients'),
                        DecorateInfo(unittest.skip("Skipped!"), 'TestJit'),
+                       DecorateInfo(unittest.skip("Skipped!"), 'TestNormalizeOperators'),
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:14029]: Normalize is probably not a redundant test?

**Review by ngimel — APPROVED** [review]
Those are cool improvements and simplifications!

File: test/test_jit.py [hunk 4]
```diff
@@ -16232,59 +16232,6 @@ def test_nhwc_autocast_jit_trace_model(model, x):
 M = 10
 S = 5
 
-
-def add_nn_functional_test(name, self_size, args, variant_name='', check_ad=(), skipTestIf=(),
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_jit.py:16236]: It has been for some time

File: test/test_ops.py [hunk 5]
```diff
@@ -139,51 +127,71 @@ def unsupported(dtype):
...
+        claimed_forward = set(op.supported_dtypes(device_type))
+        supported_but_unclaimed_forward = supported_forward - claimed_forward
+        claimed_but_unsupported_forward = claimed_forward & unsupported_forward
+
+        claimed_backward = set(op.supported_backward_dtypes(device_type))
+        supported_but_unclaimed_backward = supported_backward - claimed_backward
+        claimed_but_unsupported_backward = claimed_backward & unsupported_backward
+
+        # Partially supporting a dtype is not an error, but we print a warning
+        if (len(partially_supported_forward) + len(partially_supported_backward)) > 0:
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_ops.py:155]: Yep! And the error (and warning) messages are now much clearer

File: test/test_ops_gradients.py [hunk 6]
```diff
@@ -49,25 +50,56 @@ def is_inplace(variant):
...
-            # inputs of size n as n single Tensor inputs to gradcheck and wrap the op
-            # in a function that puts the n Tensor inputs back into a TensorList
+            # Gradcheck expects tensors as its input, but autograd actually supports tensorlists
+            #   and tensors passed as kwargs. The following creates a function that accepts just
+            #   the tensors that require grad as varargs, and then recomposes them back into the
+            #   original input.
+
+            # Creates gradcheck inputs by identifying tensors requiring grad
+            all_args = None
+            if is_iterable_of_tensors(sample.input):
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_ops_gradients.py:60]: Yes

File: test/test_ops_gradients.py [hunk 7]
```diff
@@ -49,25 +50,56 @@ def is_inplace(variant):
...
+                all_args = chain(sample.input, sample.args, sample.kwargs.values())
+            else:
+                all_args = tuple(chain((sample.input,), sample.args, sample.kwargs.values()))
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_ops_gradients.py:63]: No

File: test/test_unary_ufuncs.py (hunk 1 above)
Comment (mruberry) [reply test/test_unary_ufuncs.py:504]: We don't check safe casting to a higher dtype in test_out yet, no

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:9534]: As best the test suite can tell: yes

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:14029]: Fair point

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (ngimel) [reply torch/testing/_internal/common_methods_invocations.py:9540]: Nice. Is there a hi pri issue about this, or we don't care about `cov` (or whatever this function is)?

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -12073,8 +12073,10 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                # Pre-existing condition; Needs to be fixed
                DecorateInfo(unittest.expectedFailure, 'TestCompositeCompliance', 'test_operator', device_type='cpu'),
                # RuntimeError: "max_pool1d_impl" not implemented for 'BFloat16'
-               DecorateInfo(unittest.skip("Works on some configs"), 'TestNNCOpInfo', 'test_nnc_correctness', dtypes=(torch.bfloat16,)),
-               DecorateInfo(unittest.expectedFailure, 'TestCudaFuserOpInfo', 'test_nvfuser_correctness', dtypes=(torch.bfloat16,)),
+               DecorateInfo(unittest.skip("Works on some configs"), 'TestNNCOpInfo',
+                            'test_nnc_correctness', dtypes=(torch.bfloat16,)),
+               DecorateInfo(unittest.skip("Works on some conifgs"), 'TestCudaFuserOpInfo',
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:12078]: haha, probably on configs that are not supported by nvfuser

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -8777,6 +8777,14 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                                                   torch.bfloat16: 1e-1,
                                                   torch.complex64: 1e-2}),),
                    skips=(
+                       # Failing with wrong imaginary sign on at least some Windows jobs
+                       DecorateInfo(unittest.skip("Skipped!"), 'TestUnaryUfuncs', 'test_reference_numerics_normal',
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:8781]: this is a year old issue https://github.com/pytorch/pytorch/issues/52299, it's weird that it's still not fixed, are our windows jobs using some old cuda version?

**Pull request closed without merging**